libraries/Microsoft.Bot.Builder.Core.State/IStateStorageProvider.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder.Core.State
+{
+    public interface IStateStorageProvider
+    {
+        IStateStorageEntry CreateNewEntry(string stateNamespace, string key);
+
+        /// <summary>
+        /// Loads all state entries under the specified <paramref name="stateNamespace">partition</paramref>.
+        /// </summary>
+        /// <param name="stateNamespace"></param>
+        /// <returns></returns>
+        Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace);
```

Comment:
>Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace); [](start = 8, length = 66)

I can't change the return type for an overload on the JS side so this should be renamed to something like LoadAll()

Reply:
So this is about offering the _ability_ for the provider to optimize around the bulk loading scenario in some way, however the provider _might not_ be able to offer any sort of real optimizations around it. 

For example, the file system implementation essentially has to enumerate the files in the namespace folder on the file system and load each one. However, the Azure Table Storage implementation can obviously optimize this by doing one partition key query. Fundamentally I think you have to offer it as a separate concept at the

Reply:
> I can't change the return type for an overload on the JS side so this should be renamed to something like LoadAll()

I can certainly do that if we feel strongly enough. I feel like this could be one of those cases where we employ the patterns that are available to/established to each ecosystem. In .NET the overload makes some sense, but... again, I'm not married to it and could just change it to a `LoadAll` if you guys think that's better.

Reply:
Just checked and Redis has no out of the box support for Tags so how would you expect Redis to implement this?

---
In reply to: [179852818](https://github.com/Microsoft/botbuilder-dotnet/pull/376#discussion_r179852818) [](ancestors = 179852818,179845861)

Reply:
Replied to internal email with more details, but to have the discussion in the open there are options such as a lexing scheme that would allow us to use `ZRANGEBYLEX` or storing a list of keys for a namespace as one redis entry and then using that list to efficiently query out all the other entries in bulk.